gtsfm/frontend/correspondence_generator/det_desc_correspondence_generator.py

```diff
@@ -2,65 +2,116 @@
-                descriptors_i1_graph=delayed_descriptors[i1],
-                descriptors_i2_graph=delayed_descriptors[i2],
-                im_shape_i1=delayed_image_shapes[i1],
-                im_shape_i2=delayed_image_shapes[i2],
+
+        def apply_det_desc(det_desc: DetectorDescriptorBase, image: Image) -> Tuple[Keypoints, np.ndarray]:
+            return det_desc.detect_and_describe(image)
+
+        def apply_matcher_and_two_view_estimator(
+            feature_matcher: MatcherBase,
+            two_view_estimator: TwoViewEstimator,
+            features_i1: Tuple[Keypoints, np.ndarray],
+            features_i2: Tuple[Keypoints, np.ndarray],
+            im_shape_i1: Tuple[int, int, int],
+            im_shape_i2: Tuple[int, int, int],
+            camera_intrinsics_i1: CALIBRATION_TYPE,
+            camera_intrinsics_i2: CALIBRATION_TYPE,
+            i2Ti1_prior: Optional[PosePrior],
+            gt_camera_i1: Optional[CAMERA_TYPE],
+            gt_camera_i2: Optional[CAMERA_TYPE],
+            gt_scene_mesh: Optional[Any] = None,
+        ) -> TWO_VIEW_OUTPUT:
+            putative_corr_idxs = feature_matcher.match(
+                features_i1[0], features_i2[0], features_i1[1], features_i2[1], im_shape_i1, im_shape_i2
             )
 
-        return delayed_keypoints, delayed_putative_corr_idxs_dict
+            return two_view_estimator.run_2view(
+                features_i1[0],
+                features_i2[0],
+                putative_corr_idxs,
+                camera_intrinsics_i1,
+                camera_intrinsics_i2,
+                i2Ti1_prior,
+                gt_camera_i1,
+                gt_camera_i2,
+                gt_scene_mesh,
+            )
+
+        det_desc_future = client.scatter(self._detector_descriptor, broadcast=False)
+        feature_matcher_future = client.scatter(self._matcher, broadcast=False)
+        two_view_estimator_future = client.scatter(two_view_estimator, broadcast=False)
+        features_futures = [client.submit(apply_det_desc, det_desc_future, image) for image in images]
+
+        two_view_output_futures = {
+            (i1, i2): client.submit(
+                apply_matcher_and_two_view_estimator,
+                feature_matcher_future,
+                two_view_estimator_future,
+                features_futures[i1],
+                features_futures[i2],
+                images[i1].shape,
+                images[i2].shape,
+                camera_intrinsics[i1],
+                camera_intrinsics[i2],
+                relative_pose_priors.get((i1, i2)),
+                gt_cameras[i1],
+                gt_cameras[i2],
+                gt_scene_mesh,
+            )
+            for (i1, i2) in image_pairs
+        }
+
+        two_view_output_dict = client.gather(two_view_output_futures)
+        keypoints_futures = [client.submit(lambda f: f[0], f) for f in features_futures]
+        keypoints_list = client.gather(keypoints_futures)
```

Comment:
nit: this could go before the "apply two_view_estimator".

Reply:
I am doing this at the end as this future is already available after two view future is evaluated. This might allow dask to optimize the computations.